// tondev/dist/controllers/tonos-cli/components.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.components = void 0;
const path = __importStar(require("path"));
const core_1 = require("../../core");
function tonoscliHome() {
    return path.resolve(core_1.tondevHome(), "tonos-cli");
}
exports.components = {
    tonoscli: new class extends core_1.Component {
        getSourceName(version) {
            return `tonos-cli-${version.split(".").join("_")}-{p}.zip`;
        }
    }(tonoscliHome(), "tonos-cli", {
        resolveVersionRegExp: /tonos_cli\s+([0-9.]+)/,
        executable: true,
        globally: true,
    }),
};
