// tondev/dist/controllers/solidity/components.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.components = void 0;
const path_1 = __importDefault(require("path"));
const core_1 = require("../../core");
function solidityHome() {
    return path_1.default.resolve(core_1.tondevHome(), "solidity");
}
exports.components = {
    compiler: new core_1.Component(solidityHome(), "solc", {
        executable: true,
    }),
    linker: new core_1.Component(solidityHome(), "tvm_linker", {
        executable: true,
        resolveVersionRegExp: /[^0-9]*([0-9.]+)/,
    }),
    stdlib: new class extends core_1.Component {
        getSourceName(version) {
            return `${this.name}_${version.split(".").join("_")}.tvm.gz`;
        }
        async resolveVersion(downloadedVersion) {
            return downloadedVersion;
        }
        async loadAvailableVersions() {
            return exports.components.compiler.loadAvailableVersions();
        }
    }(solidityHome(), "stdlib_sol", {
        targetName: "stdlib_sol.tvm",
    }),
};
