// tondev/dist/controllers/network/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkTool = exports.networkGiverCommand = exports.networkDefaultCommand = exports.networkDeleteCommand = exports.networkInfoCommand = exports.networkListCommand = exports.networkAddCommand = void 0;
const registry_1 = require("./registry");
const utils_1 = require("../../core/utils");
const forceArg = {
    name: "force",
    alias: "f",
    type: "boolean",
    title: "Overwrite key if already exists",
    defaultValue: "false",
};
const nameArg = {
    isArg: true,
    name: "name",
    type: "string",
    title: "Network name",
};
exports.networkAddCommand = {
    name: "add",
    title: "Add net",
    args: [
        nameArg,
        {
            isArg: true,
            name: "endpoints",
            type: "string",
            title: "Comma separated endpoints",
        },
        forceArg,
    ],
    async run(_terminal, args) {
        const endpoints = args.endpoints.split(",").filter(x => x !== "");
        new registry_1.NetworkRegistry().add(args.name, "", endpoints, args.force);
    },
};
exports.networkListCommand = {
    name: "list",
    alias: "l",
    title: "Prints list of networks",
    args: [],
    async run(terminal, _args) {
        const registry = new registry_1.NetworkRegistry();
        const rows = [["Network", "Endpoints", "Giver", "Description"]];
        registry.items.forEach((network) => {
            var _a, _b, _c;
            rows.push([
                `${network.name}${network.name === registry.default ? " (Default)" : ""}`,
                registry_1.NetworkRegistry.getEndpointsSummary(network),
                (_b = (_a = network.giver) === null || _a === void 0 ? void 0 : _a.address) !== null && _b !== void 0 ? _b : "",
                (_c = network.description) !== null && _c !== void 0 ? _c : "",
            ]);
        });
        const table = utils_1.formatTable(rows, { headerSeparator: true });
        if (table.trim() !== "") {
            terminal.log(table);
        }
    },
};
exports.networkInfoCommand = {
    name: "info",
    alias: "i",
    title: "Prints network detailed information",
    args: [
        {
            ...nameArg,
            defaultValue: "",
        },
    ],
    async run(terminal, args) {
        if (args.name === "") {
            return exports.networkListCommand.run(terminal, {});
        }
        const registry = new registry_1.NetworkRegistry();
        const network = registry.get(args.name);
        const rows = [["Network", network.name]];
        rows.push(["Endpoints", network.endpoints.join(", ")]);
        const giver = network.giver;
        if (giver) {
            rows.push([
                "Giver",
                `${giver.name} at ${giver.address}${giver.signer ? ` signed by ${giver.signer}` : ""}`,
            ]);
        }
        if (network.name === registry.default) {
            rows.push(["Default", "true"]);
        }
        if (network.description) {
            rows.push(["Description", network.description]);
        }
        terminal.log(utils_1.formatTable(rows));
    },
};
exports.networkDeleteCommand = {
    name: "delete",
    title: "Delete network from registry",
    args: [nameArg],
    async run(_terminal, args) {
        new registry_1.NetworkRegistry().delete(args.name);
    },
};
exports.networkDefaultCommand = {
    name: "default",
    alias: "d",
    title: "Set default network",
    args: [nameArg],
    async run(_terminal, args) {
        new registry_1.NetworkRegistry().setDefault(args.name);
    },
};
exports.networkGiverCommand = {
    name: "giver",
    alias: "g",
    title: "Set giver for network",
    args: [
        nameArg,
        {
            isArg: true,
            name: "address",
            title: "Giver address",
            type: "string",
            defaultValue: "",
        },
        {
            name: "signer",
            alias: "s",
            title: "Signer to be used with giver",
            type: "string",
            defaultValue: "",
        },
        {
            name: "value",
            alias: "v",
            title: "Deploying account initial balance in nanotokens",
            type: "string",
            defaultValue: "",
        },
    ],
    async run(_terminal, args) {
        const value = utils_1.parseNumber(args.value);
        return new registry_1.NetworkRegistry().setGiver(args.name, args.address, args.signer, value);
    },
};
exports.NetworkTool = {
    name: "network",
    alias: "n",
    title: "Network Registry",
    commands: [
        exports.networkAddCommand,
        exports.networkDeleteCommand,
        exports.networkListCommand,
        exports.networkInfoCommand,
        exports.networkDefaultCommand,
        exports.networkGiverCommand,
    ],
};
