// tondev/dist/controllers/clang/components.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.components = void 0;
const os = __importStar(require("os"));
const path_1 = __importDefault(require("path"));
const core_1 = require("../../core");
const p = os.platform();
let innerPath;
let targetName;
let ext;
if (p === "linux") {
    innerPath = "opt/work/llvm/install/bin/clang";
    targetName = "clang.tar";
    ext = p + ".tar.gz";
}
else if (p === "darwin") {
    innerPath = "bin/clang";
    targetName = "clang.zip";
    ext = p + ".zip";
}
else {
    innerPath = "bin/clang.exe";
    targetName = "clang.zip";
    ext = p + ".zip";
}
function clangHome() {
    return path_1.default.resolve(core_1.tondevHome(), "clang");
}
exports.components = {
    clang: new (class extends core_1.Component {
        getSourceName(version) {
            return `${this.name}/${this.name}-${version.split(".").join("_")}-${ext}`;
        }
        async resolveVersion(downloadedVersion) {
            return downloadedVersion;
        }
    })(clangHome(), "clang-for-tvm", {
        targetName,
        innerPath,
        executable: true,
    }),
};
